net: add autoSelectFamily option (#44731)
nodejs/node: doc/api/net.md

[`'connection'`][] event emitted on a [`net.Server`][], so the user can use
it to interact with the client.

> ### Class property: `net.Socket.autoDetectFamily`

Reviewer: I'm wary of non-`const` global properties that affect the entire application, e.g., most such properties turned out to be a mistake in `node:crypto`.

Reviewer 2: +1 on this feedback. It's hard for me to think of a place where we've done this and it hasn't been an unwelcome foot-gun.

Author: I'm not fan of this as well actually.
I had to introduce otherwise most tests in the CI would fail as they use the `connection` event and `common.mustCall`. Which will fail with happy eyeballs attempts two connections.

Do you have a better idea?

Reviewer 3: I actually like this. It's a one-line-fix for most IPv6 issues if we decide to not enable happy eyeballs by default.

Author: So shall I leave it or not?

Reviewer: @mcollina Wouldn't an environment variable make more sense then if your intention is to quickly switch it on for an existing application? A global property requires changes to the program code and it cannot be set by libraries because that could break other components (which is why I am against global properties).

Reviewer: (Environment variables are jut as bad as global variables, of course, and it makes it more difficult to change the behavior from within the application.)

Author: Yes, I can definitely go for an environment variable if it is an approach we accept.

* `port` {number} Required. Port the socket should connect to.
* `host` {string} Host the socket should connect to. **Default:** `'localhost'`.
* `localAddress` {string} Local address the socket should connect from.

>   This is ignored if `autoDetectFamily` is set to `true`.

Reviewer: Does enabling it by default break applications that rely on `localAddress` etc.? The documentation says that the option will be ignored when `autoDetectFamily` is enabled.

Reviewer 4: I feel like it should be the opposite: ignore `autoDetectFamily` if `localAddress` is specified. Unless there's another solution that doesn't make these options mutually exclusive.

Reviewer: Yes, I don't like ignoring pre-existing options either.

Reviewer 3: Agreed, we should ignore `autoDetectFamily` if `localAddres` is set.

Author: Changing it.

  **Default:** `false`.
* `keepAliveInitialDelay` {number} If set to a positive number, it sets the initial delay before
  the first keepalive probe is sent on an idle socket.**Default:** `0`.

> * `autoSelectFamily` {boolean|number}: If set to `true` or to a positive number, it enables

Reviewer: It feels strange to mix `boolean|number` here instead of splitting it into two options. The meaning of the `number` is also not obvious and the reader must read the entire paragraph to understand its meaning.

Author: You are right. It's confusing. I'll split in `autoSelectFamily` and `autoSelectFamilyAttemptTimeout`.

  a family autodetection algorithm that loosely implements section 5 of [RFC 8305][]. The `all`
  option passed to lookup is set to `true` and the sockets attempts to connect to all

>   returned AAAA and A records, in sequence, until a successful connection is established.

Reviewer: ```suggestion
  obtained IPv6 and IPv4 addresses, in sequence, until a connection is established.
```

Author: 👍

  The first returned AAAA record is tried first, then the first returned A record and so on.

>   The option specifies how much time in milliseconds to wait for a attempt to finish before trying

Reviewer: I don't think `finish` is the right word here.

```suggestion
  The option specifies how much time in milliseconds to wait for an attempt to finish before trying
```

Author: 👍

  the next one.

>   Connection errors are not emitted if at least a connection succeeds.

Reviewer: ```suggestion
  Connection errors are not emitted if at least one connection succeeds.
```

Author: 👍

  If set to `true`, then the value `250` will be used instead.
  If set to a positive integer less than `10`, then the value `10` will be used instead.
  Ignored if the `family` option is not `0` or if `localAddress` is set.

>   **Default:** `0`.

Reviewer: This seems misleading at best.

```suggestion
  **Default:** false.
```

Author: Superseded by option change.

  **Default:** `false`.
* `keepAliveInitialDelay` {number} If set to a positive number, it sets the initial delay before
  the first keepalive probe is sent on an idle socket.**Default:** `0`.
* `autoSelectFamily` {boolean}: If set to `true`, it enables a family autodetection algorithm
  that loosely implements section 5 of [RFC 8305][].
  The `all` option passed to lookup is set to `true` and the sockets attempts to connect to all
  obtained IPv6 and IPv4 addresses, in sequence, until a connection is established.

>   The first returned AAAA address is tried first, then the first returned A address and so on.

Reviewer 5: This sentence is a bit vague. If multiple `AAAA` addresses are returned and multiple `A` addresses are returned, is the sequence tried `AAAA`, then `A`, then `AAAA`, ... or does it try all of the `AAAA` addresses first then all of the `A`?

  Each connection attempt is given the amount of time specified by the `autoSelectFamilyAttemptTimeout`
  option before timing out and trying the next address.
  Ignored if the `family` option is not `0` or if `localAddress` is set.

>   Connection errors are not emitted if at least one connection succeeds.

Reviewer 5: For debugging purposes, one thing to consider adding *later* (not in this PR) is a property that provides a listing of the addresses that were unsuccessfully tried.

  **Default:** `false`.
* `keepAliveInitialDelay` {number} If set to a positive number, it sets the initial delay before
  the first keepalive probe is sent on an idle socket.**Default:** `0`.
* `autoDetectFamily` {boolean}: Enables the Happy Eyeballs connection algorithm.
  The `all` option passed to lookup is set to `true` and the sockets attempts to
  connect to all returned AAAA and A records at the same time, keeping only
  the first successful connection and disconnecting all the other ones.
  Connection errors are not emitted if at least a connection succeeds.

>   Ignored if the `family` option is not `0`.

Reviewer 3: Please write what the default value is. This should be `true`.

  **Default:** `false`.
* `keepAliveInitialDelay` {number} If set to a positive number, it sets the initial delay before
  the first keepalive probe is sent on an idle socket.**Default:** `0`.

> * `autoDetectFamily` {boolean}: Enables the Happy Eyeballs connection algorithm.

Reviewer: Assuming all addresses are functioning, it's not so much about _detecting_ address families, but more about selecting the fastest one. How about `autoSelectFamily`, `autoFastestAddress`, or something in that direction? (I am not good at coming up with names right now.)

[IPC]: #ipc-support
[Identifying paths for IPC connections]: #identifying-paths-for-ipc-connections

> [RFC 8305]: https://www.rfc-editor.org/rfc/rfc8305.txt

Reviewer 6: [](url)

> [RFC 8305]: https://www.rfc-editor.org/rfc/rfc8305.txt

Reviewer 6: Duplicate of #
